<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get label for inputs — get_inputs_label • tidycwl</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Get label for inputs — get_inputs_label"><meta property="og:description" content="Get label for inputs"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidycwl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/tidycwl.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/sbg/tidycwl/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get label for inputs</h1>
    <small class="dont-index">Source: <a href="https://github.com/sbg/tidycwl/blob/HEAD/R/get-param.R" class="external-link"><code>R/get-param.R</code></a></small>
    <div class="hidden name"><code>get_inputs_label.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get label for inputs</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_inputs_label</span><span class="op">(</span><span class="va">inputs</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>inputs</dt>
<dd><p>Parsed inputs</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Vector of input labels</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># inputs represented by a dictionary</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"cwl/sbg/workflow/rnaseq-salmon.json"</span>, package <span class="op">=</span> <span class="st">"tidycwl"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="read_cwl_json.html">read_cwl_json</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="parse_inputs.html">parse_inputs</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">get_inputs_label</span><span class="op">(</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "Transcriptome FASTA or Salmon index archive"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2] "FASTQ Read Files"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "Gene map or GTF file"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "Maximum number of parallel jobs"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "K-mer length"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "GENCODE FASTA"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "Keep duplicates"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "Write unmapped names"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "Write mappings"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "VBEM prior"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "Use Variational Bayesian optimization"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12] "Thinning factor"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "Strict intersect"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14] "Sequence-specific bias correction"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15] "Reduce GC memory"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16] "Range factorization bins"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "Quasi coverage"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18] "Position bias"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] "Per transcript prior"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [20] "Number of pre auxiliary model samples"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21] "Number of Gibbs samples"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [22] "Number of bootstraps"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [23] "Number of bias samples"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [24] "Number of auxiliary model samples"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "No length correction"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [26] "No fragment length distribution"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [27] "No effective length correction"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [28] "No bias length threshold"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29] "Minimum assigned fragments"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [30] "Meta"                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] "Maximum read occurence"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [32] "Maximum (S)MEM occurance"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [33] "Initialize uniform parameters"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [34] "Incompatible prior probability"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [35] "GC size sample"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [36] "GC bias correction"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] "Forgetting factor"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [38] "Fragment length standard deviation"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [39] "Mean fragment length"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [40] "Maximum fragment length"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [41] "Faster mapping"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [42] "Dump equivalence class weights"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] "Dump equivalence class counts"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [44] "Discard orphans in Quasi-mapping mode"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [45] "Consistent hits"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [46] "Bias speed sample"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [47] "Alternative initialization mode"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [48] "Allow orphans in FMD mode"                  </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># inputs represented by a list</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"cwl/sbg/workflow/rnaseq-salmon.cwl"</span>, package <span class="op">=</span> <span class="st">"tidycwl"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="read_cwl_yaml.html">read_cwl_yaml</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="parse_inputs.html">parse_inputs</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">get_inputs_label</span><span class="op">(</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "Transcriptome FASTA or Salmon index archive"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2] "FASTQ Read Files"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "Gene map or GTF file"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "Maximum number of parallel jobs"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "K-mer length"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "GENCODE FASTA"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "Keep duplicates"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "Write unmapped names"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "Write mappings"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "VBEM prior"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "Use Variational Bayesian optimization"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12] "Thinning factor"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "Strict intersect"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14] "Sequence-specific bias correction"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15] "Reduce GC memory"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16] "Range factorization bins"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "Quasi coverage"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18] "Position bias"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] "Per transcript prior"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [20] "Number of pre auxiliary model samples"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21] "Number of Gibbs samples"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [22] "Number of bootstraps"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [23] "Number of bias samples"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [24] "Number of auxiliary model samples"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "No length correction"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [26] "No fragment length distribution"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [27] "No effective length correction"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [28] "No bias length threshold"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29] "Minimum assigned fragments"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [30] "Meta"                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] "Maximum read occurence"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [32] "Maximum (S)MEM occurance"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [33] "Initialize uniform parameters"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [34] "Incompatible prior probability"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [35] "GC size sample"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [36] "GC bias correction"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] "Forgetting factor"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [38] "Fragment length standard deviation"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [39] "Mean fragment length"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [40] "Maximum fragment length"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [41] "Faster mapping"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [42] "Dump equivalence class weights"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] "Dump equivalence class counts"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [44] "Discard orphans in Quasi-mapping mode"      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [45] "Consistent hits"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [46] "Bias speed sample"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [47] "Alternative initialization mode"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [48] "Allow orphans in FMD mode"                  </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jeffrey Grover, Soner Koc, Nan Xiao, Dennis Dean, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

